session_sync.go

```diff
@@ -0,0 +1,82 @@
+package remotedialer
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"time"
+)
+
+var errCloseSyncConnections = errors.New("sync from client")
+
+// encodeConnectionIDs serializes a slice of connection IDs
+func encodeConnectionIDs(ids []int64) []byte {
+	payload := make([]byte, 0, 8*len(ids))
```

Comment:
nit-pick and open question for reviewers:

The size of the payload will be proportional to the number of active connections and depend on the type used for this value (`int64` at the moment, which needs 8 bytes to represent up to `9,22e18` (19 digits) positive values, since the counter starts at 0).
We could halve the payload size by using `uint32`, which would still leave us more than 4 billion unique ids to use, which I believe is more than enough for this purpose.
In fact, these ID are used in the context of a session (websocket multiplexing), so we could go even further and use `uint16`'s `65536` unique values.
The current implementation relies on `int64` providing an "almost infinite" number of IDs so they are not repeated, but we could also prevent that using a smaller size by ensuring a new connection's ID is not included in the connections map.

Finally, please beware this won't only affect the payload size, but also reduce the overall memory usage by remotedialer (although this may represent a tiny percentage from total Rancher usage).

Reply:
Do we have a ballpark idea of how many active connections are opened in a typical case and in big clusters? Are we talking about 1000s, 10000s, more?

Reply:
@tomleb Currently on a rancher instance with around 3.4k connected downstream clusters I see 66.5k connections.
```
sum(session_server_total_add_connections - session_server_total_remove_connections)
```

The main question in general to me seems: What is the number of downstream cluster which Rancher should support and is remotedialer able to handle this amount of clusters/traffic?
Currently in our case we're not able to scale rancher to more than one pod as it seems that the remotedialer peer connections are not able to handle the load (see https://github.com/rancher/rancher/issues/45129). With a single pod we don't have peer connections and hence its working more or less.

Reply:
@moio thank you for the response.

>  I believe you are affected by the problem this PR is trying to fix

I didn't look into this PR in detail. But should the number of connections not grow steadily as they get lost?
Because in our case the connection count only goes up until all clusters are active. After that they stay on around 66.5k.
![number_of_connections](https://github.com/rancher/remotedialer/assets/19696698/8dadbd17-fa6f-4fcc-9e6d-b720aa22aa29)

Reply:
Its repeatedly ramping up because rancher is restarting all the time (see https://github.com/rancher/rancher/issues/45164). But during off-peak hours it usually stays steady on 66.5k for longer periods of time without restarting.

From my understanding the ramp up is because it takes some time until all controllers for the downstream clusters are started (user controllers?). After that all connected downstream clusters are in active state. Before they are in unavailable state. We see the same ramp up pattern for memory and network.
My assumtion is that during this startup phase the controllers are opening many connections because they are starting listeners (LIST WATCH)?

Reply:
> Say we'd use int64 then that'd be around 8KB sent every minute. Doesn't sound to bad to me, though maybe others disagree.

I agree, probably not worth it.

> If we plan on optimizing this in the future, it could help adding an "encoding type" byte to the sync connection payload so we can keep backwards compatible with previous version of remotedialer (eg: when doing a rolling upgrade). Another way would be to create a new SyncConnection message type, but the encoding type seems simpler to me.

That's a good idea. After thinking about it, maybe we don't even need to be this careful. If we ever changed it, all the remotedialer peers would be running the same version, since they would have the same Rancher image.
The only exception would be the upgrade time window, but we can basically ignore that, because all those connections are going to be dropped anyways.

---
In summary, I think there is no action needed here, so I'm going to resolve this conversation. Thanks for participating!